src/libstd/io/lazy.rs

```diff
@@ -1,64 +1,77 @@
-use cell::Cell;
-use ptr;
-use sync::Arc;
+use cell::UnsafeCell;
+use sync::atomic::AtomicUsize;
+use sync::atomic::Ordering;
 use sys_common;
 use sys_common::mutex::Mutex;
 
+/// Helper for lazy initialization of a static, with a destructor that runs when the main (Rust)
+/// thread exits.
+///
+/// Currently used only inside the standard library, by the stdio types.
+///
+/// # Safety
+/// - `UnsafeCell`: We only create a mutable reference during initialization and during the shutdown
+///   phase. At both times there can't exist any other references.
+/// - Destruction. The `Drop` implementation of `T` should not access references to anything except
+///   itself, they are not guaranteed to exist. It should also not rely on other machinery of the
+///   standard library to be available.
+/// - Initialization. The `init` function for `get` should not call `get` itself,  to prevent
+///   infinite recursion and acquiring the guard mutex reentrantly.
+/// - We use the `Mutex` from `sys::common` because it has a `const` constructor. It currently has
+///   UB when acquired reentrantly without calling `init`.
```

Comment:
Leave a FIXME for the UB?

Reply:
It should be gone when the parking_lot PR lands, which also touches this part of the code. I will keep an eye on it and remove the comment if it lands first. cc @faern.

Note that there is no UB here as we don't use it reentrantly just like the old code. It is more a general comment to explain the parts this code is careful with in the `unsafe` block.